[#75] 공연 스케줄, 잔여 좌석 조회

tlsgkr7416 opened the pull request:
### 기능 구현

#### 1. 스케줄 조회
- controller : 공연 id를 받아 스케줄 조회

- service
    * 공연 시간과 제목을 조회
    * 공연 id가 존재하는지 체크
    
- mapper
    * join을 통해 공연 스케줄과 제목을 불러옴
    * 공연 id가 존재하는지 체크
- 기능
    *  공연 스케줄을 조회 시 스케줄이 없어도 빈 배열로 조회
    *  스케줄을 조회 시 프론트에서는 달력을 만들어 user에게 조회
    *  예매 후 일반 회원만 이용이 가능하도록 적용
  
#### 2. 잔여 좌석 조회
- controller : 공연 id와 공연 시간 id를 요청 시 공연 시작 시간, 공연 시간Id, 구매 가능한 좌석 수를 응답
- service
    
    * 공연 시간, 구매 가능한 좌석 수를 응답하는 서비스
    * 공연 id가 존재하는지 체크
    * 공연 날짜가 존재하는지 체크
- mapper
   
    * 공연 시간과 구매 가능한 좌석을 응답
    * 공연 날짜가 존재하는지 체크
    * 공연 id가 존재하는지 체크
- 기능
   
    * 공연 예매를 클릭 후 스케줄을 보여주고 스케줄을 클릭 시 구매 가능한 좌석을 조회하는 기능
    * 예매 후 일반 회원만 이용이 가능하도록 적용

#### 3. 캐시 기능 구현

- PERFORMANCE_TIME : 공연 상세페이지에서 예매를 클릭 후 스케줄을 선택하는 화면으로 시간이 추가 될 가능성을 고려해 5분으로 적용
- PERFORMANCE_SEAT : 구매 가능한 공연 좌석을 보여주는 화면으로 인기 있는 공연일 경우 좌석 수가 빠르게 업데이트될 가능성을 고려해 3초로 적용 (1000명이 동시에 구매 가능한 좌석 데이터를 보여줄 때 케시에 있는 데이터를 보여주면 효율적이여서 3초라도 캐시를 적용)
---
### 기능 리스트

- [x]  공연 스케줄 조회 기능 구현
- [x]  공연 좌석 조회 기능 구현
- [x]  캐시 기능 구현
- [x]  테스트 코드 작성

tlsgkr7416 committed:
[#75] 공연 스케줄 조회 기능 구현

- controller : 공연 id를 받아 스케줄 조회
- service
    * 공연 시간과 제목을 조회
    * 공연 좌석, 스케줄이 없을 경우 예매를 못하게 처리
- mapper
    * inner join을 통해 공연 스케줄과 제목을 불러옴
    * 공연, 공연 시간, 좌석 테이블을 inner join 하여 좌석이 예약되었거나 좌석이 없을 경우 체크

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/feature/98/sh' into feature/75/sh

tlsgkr7416 committed:
[#75] 공연 스케줄 조회 advice 기능 추가

f-lab-dev commented on src/main/java/com/show/showticketingservice/controller/PerformanceController.java:
```diff
@@ -73,4 +69,9 @@ public void deletePerformance(@PathVariable int performanceId) {
         performanceService.deletePerformance(performanceId);
     }
 
+    @GetMapping("/{performanceId}/times")
```
👍

tlsgkr7416 committed:
[#75] 공연 좌석 조회 기능 구현, 캐시 기능 구현

-PerformanceController : 공연 id와 공연 날짜를 요청 시 공연 시작 시간, 공연 시간Id, 구매 가능한 좌석 수를 응답
-PerformanceService
    * getPerfTimeAndSeatCapacity : 공연 시간, 구매 가능한 좌석을 응답하는 서비스
    * checkValidPerformanceId : 공연 id가 존재하는지 체크
    * checkPerfDate : 공연 날짜가 존재하는지 체크
-performanceTimeMapper
    * getPerfTimeAndSeatCapacity : 공연 시간과 구매 가능한 좌석을 응답하는 mapper
    * isPerfDate : 공연 날짜가 존재하는지 체크하는 mapper
-캐시
    * PERFORMANCE_TIME : 공연 상세페이지에서 예매를 클릭 후 스케줄을 선택하는 화면으로 인기 있는 공연일 경우 빨리 매진될 가능성을 고려해 5분으로 적용
    * PERFORMANCE_SEAT : 구매 가능한 공연 좌석을 보여주는 화면으로 인기 있는 공연일 경우 좌석 수가 빠르게 업데이트될 가능성을 고려해 3초로 적용 (1000명이 동시에
      구매 가능한 좌석 데이터를 보여줄 때 케시에 있는 데이터를 보여주면 효율적이여서 3초라도 캐시를 적용)
-기능
    * 공연 예매를 클릭 후 스케줄을 보여주고 스케줄을 클릭 시 구매 가능한 좌석을 조회하는 기능
    * 예매 후 일반 회원만 이용이 가능하도록 적용

hanwix2 commented on src/main/java/com/show/showticketingservice/controller/PerformanceController.java:
```diff
@@ -73,4 +69,9 @@ public void deletePerformance(@PathVariable int performanceId) {
 
+    @GetMapping("/{performanceId}/times")
+    @UserAuthenticationNecessary
+    public List<PerformanceTitleAndTimesResponse> getPerformanceTitleAndTimes(@PathVariable int performanceId) {
```
List를 반환하는 이유가 궁금합니다.

`PerformanceTitleAndTimesResponse` 의 인스턴스로 이미 **공연 시간 List**가 존재하기도 하고
**id**와 **title**이 불필요하게 중복되지 않을까요?

tlsgkr7416 replied:
여기는 제가 실수했네요. 테스트 코드 작성하다가 저도 발견해서 수정했어요!

hanwix2 commented on src/main/java/com/show/showticketingservice/model/performance/PerformanceTimeResponse.java:
```diff
@@ -8,6 +8,8 @@
 @NoArgsConstructor(access = AccessLevel.PRIVATE)
 public class PerformanceTimeResponse {
 
+    private int id;
```
**기존의 DTO 클래스를 수정**하는 경우 이것을 사용하는 다른 로직에 영향이 갈 수 있으니 주의 깊게 확인해주시고 저에게도 작업시 알려주시면 좋을 것 같아요!

이건 현재 PR이 아닌 다른 PR에서의 작업이긴 하지만 이전에 비슷한 사례로 버그 발생 원인이었기에 중요하다고 생각해서 같이 잊지 않기 위해 기록해둡니다.
_(일단 이 작업에 대한 문제점은 없습니다 ㅎㅎ)_

tlsgkr7416 committed:
[#75] 공연 스케줄, 잔여 좌석 조회 수정, merge 충돌 해결, 테스트 코드 작성

tlsgkr7416 committed:
[#75] 공연 스케줄 조회 수정

공연 스케줄 조회 기능에서 프론트에서는 달력을 만들 때 시간은 필요 없기 때문에 날짜만 보내도록 수정

tlsgkr7416 committed:
[#75] 공연 스케줄, 좌석 조회 주석 수정

tlsgkr7416 reviewed (commented):
'[#75] 공연 스케줄, 잔여 좌석 조회 수정, merge 충돌 해결, 테스트 코드 작성'
여기 commit은 충돌 해결 후 commit 시켜야 됐는데 까먹고 테스트 코드랑 조회 기능 약간 수정했네요. 조회 기능은 공연 스케줄에서 응답 데이터를 List만 수정했어요. 따라서 테스트 코드만 봐주시면 될 거 같아요!

hanwix2 commented on src/main/java/com/show/showticketingservice/model/performance/SeatCapacityResponse.java:
```diff
@@ -0,0 +1,15 @@
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class SeatCapacityResponse {
+
+    private RatingType ratingType;
+
+    private int seatCapacity;
```
**SeatCapacity** 라는 표현은 남은 좌석이라고 생각하기엔 나중에 총 좌석수라는 의미와 혼동이 될 수 있을 것 같은데
**RemainingSeats** 같은 표현은 어떨까요?

hanwix2 commented on src/main/java/com/show/showticketingservice/mapper/PerformanceMapper.java:
```diff
@@ -32,6 +32,10 @@
 
+    PerformanceTitleAndTimesResponse getPerformanceTitleAndTimes(int performanceId);
+
+    boolean isPerfTicket(int performanceId);
```
메소드명의 의미는 **공연티켓인지 아닌지** 여부를 뜻하는데 
**남는 자리가 있는지 체크**하는 쿼리와는 의미가 다른 것 같습니다.

hanwix2 commented on src/main/java/com/show/showticketingservice/mapper/PerformanceTimeMapper.java:
```diff
@@ -13,4 +14,8 @@
+
+    List<PerfTimeAndSeatCapacityResponse> getPerfTimeAndSeatCapacity(int performanceId, String perfDate);
+
+    boolean isPerfDate(int performanceId, String perfDate);
```
이 메소드명도 의미가 모호한 것 같습니다. 
유효한 스케줄인지의 여부를 확인한다는 의미가 확실히 드러나야 할 것 같아요.

hanwix2 commented on src/main/java/com/show/showticketingservice/controller/PerformanceController.java:
```diff
@@ -77,4 +77,16 @@ public void deletePerformance(@PathVariable int performanceId) {
+
+    @GetMapping("/{performanceId}/times/{perfDate}")
+    @UserAuthenticationNecessary(role = AccessRoles.GENERAL)
+    public List<PerfTimeAndSeatCapacityResponse> getPerfTimeAndSeatCapacity(@PathVariable int performanceId, @PathVariable String perfDate) {
+        return performanceService.getPerfTimeAndSeatCapacity(performanceId, perfDate);
```
`perfDate`를 **String**으로 받으면 처리하는 과정과 쿼리가 더 복잡해지지 않을까요?


서비스 흐름을 생각해 볼 때 잔여좌석을 요청하기 전 해당 공연의 스케줄들을 조회하고 UI에서 특정 스케줄을 선택해서 잔여좌석을 확인할텐데
스케줄의 **ID** 값을 전달하는 게 더 간편하고 안전(String으로 받으면 날짜가 의도한 형식으로 오지 않을 가능성 존재)하지 않을까요?

f-lab-dev commented on src/main/java/com/show/showticketingservice/service/PerformanceService.java:
```diff
@@ -299,16 +295,46 @@ private void checkValidPerfIdAndShowType(ShowType showType, Integer lastPerfId)
+    }
+
+    @Cacheable(cacheNames = CacheConstant.PERFORMANCE_SEAT, key = "#performanceId + #perfDate")
+    public List<PerfTimeAndSeatCapacityResponse> getPerfTimeAndSeatCapacity(int performanceId, String perfDate) {
```
날짜를 검색하는건데 `LocalDate`가 아니라 문자열로 받는 이유가 있으실까요?
문자열은 실수의 여지가 많아보입니다~

f-lab-dev commented on src/test/java/com/show/showticketingservice/service/PerformanceServiceTest.java:
```diff
@@ -94,4 +100,117 @@ public void getPerformanceListFailPageFunctionIdNotExists() {
+    }
+
+    @Test
+    @DisplayName("공연 예매 시 스케줄을 선택하면 남은 좌석 수를 조회합니다.")
```
`공연 ID와 날짜로 좌석수를 조회한다.` 라는 행위에만 집중해야합니다. 공연 예매에 종속적인 제목이네요~

tlsgkr7416 committed:
[#75] 공연 스케줄, 잔여 좌석 조회 피드백 반영

- 메소드, 클래스명 등 수정
- 공연 스케줄 조회 시 남은 좌석이 없거나 스케줄이 등록되어 있지 않을 때 예외 처리 제거
    * 남은 좌석이 없어도 공연 스케줄을 확인하고 싶을 수도 있다는 가능성을 고려
    * 스케줄이 등록되어 있지 않을 시 빈 배열로 반환
- 공연 좌석 조회 시 시간 요청 대신 공연 시간 id로 요청 변경
- 테스트 코드 수정

tlsgkr7416 committed:
[#75] 공연 스케줄, 잔여 좌석 조회 오타 수정

tlsgkr7416 committed:
Merge remote-tracking branch 'origin/develop' into feature/75/sh
